bootstrap: add quoting support to avoid splitting

kiike opened the pull request:
With this change, it is now possible to pass quotes to the configure script, such as

`./configure.py --set=target.\"thumbv8m.main-none-eabi\".linker=/linker`
or
`./configure.py '--set=target."thumbv8m.main-none-eabi".linker=/linker'`

, which will treat `thumbv8.main-none-eabi` as a whole part. Currently, the string would be split into two elements: `thumbv8`, and `main-none-eabi`.

The approach taken is to perform custom splitting instead of using `str.split()` and then repairing the split. Also, There are numerous corner cases not handled: the custom split doesn't differentiate between single quotes or double quotes, so it is perfectly possible to pass `./configure.py --set=target.\"thumbv8m.main-none-eabi\'.linker=/linker` and the behaviour would be the same as with all double quotes or single quotes.

As for the code, i'm unsure on whether to delimit strings with double or single quotes. I've seen both single quotes and double quotes used to delimit strings, like in 
```py
err("Option '{}' provided more than once".format(key))
```
and this a handful of lines down:
```py
if option.name == 'sccache':
    set('llvm.ccache', 'sccache', config)
```
Please advise on the wanted one.

Fixes #130602

r? @onur-ozkan

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

Thanks in advance for the feedback!

kiike committed:
bootstrap: add quoting support to avoid splitting

With this change, it is now possible to pass quotes to the configure
script, such as

`./configure.py --set=target.\"thumbv8m.main-none-eabi\".linker=/linker`

, which will treat `thumbv8.main-none-eabi` as a whole part. Currently,
the string would be split into two elements: `thumbv8`, and
`main-none-eabi`.

jieyouxu commented:
Can this use something like <https://docs.python.org/3.10/library/shlex.html#shlex.split>?

onur-ozkan commented:
> Can this use something like <https://docs.python.org/3.10/library/shlex.html#shlex.split>?

Does that work for python2 aswell?

jieyouxu commented:
> Does that work for python2 aswell?

AFAICT that exists for python2 https://docs.python.org/2.7/library/shlex.html as well

kiike commented:
> Can this use something like https://docs.python.org/3.10/library/shlex.html#shlex.split?

Not as is. That splits on whitespace, right? We want to split on periods. If you want me to use `shlex.split`, I need to research how to modify its behaviour to split on periods, because right now it splits on whitespace with shell semantics.

onur-ozkan commented:
Bootstrap should be compatible with python versions older than 3.6. Some distros (and possibly even some of our build runners) might still use very old python versions.

kiike commented:
> Bootstrap should be compatible with python versions older than 3.6. Some distros (and possibly even some of our build runners) might still use very old python versions.

Got it. I've added the relevant word character to the last commit.

onur-ozkan reviewed (approved):
LGTM.

Could you please squash your commits?

onur-ozkan commented:
@rustbot author

kiike commented:
@rustbot review

@onur-ozkan done!

onur-ozkan commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 8471c6bb0c3f46f5dbad5942aedb661304cae001 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 8471c6bb0c3f46f5dbad5942aedb661304cae001 -->
<!-- homu: {"type":"Approved","sha":"8471c6bb0c3f46f5dbad5942aedb661304cae001","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8471c6bb0c3f46f5dbad5942aedb661304cae001 with merge fe43131683267082fc46be5796ce84351e9fb497...
<!-- homu: {"type":"BuildStarted","head_sha":"8471c6bb0c3f46f5dbad5942aedb661304cae001","merge_sha":"fe43131683267082fc46be5796ce84351e9fb497"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11720536209/job/32655297402)
Approved by: onur-ozkan
Pushing fe43131683267082fc46be5796ce84351e9fb497 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11720536209/job/32655297402"},"merge_sha":"fe43131683267082fc46be5796ce84351e9fb497"} -->

rust-timer commented:
Finished benchmarking commit (fe43131683267082fc46be5796ce84351e9fb497): [comparison URL](https://perf.rust-lang.org/compare.html?start=9200cbc71230a86b8a70481b625d93a4fa5ef523&end=fe43131683267082fc46be5796ce84351e9fb497&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9200cbc71230a86b8a70481b625d93a4fa5ef523&end=fe43131683267082fc46be5796ce84351e9fb497&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9200cbc71230a86b8a70481b625d93a4fa5ef523&end=fe43131683267082fc46be5796ce84351e9fb497&stat=max-rss)
<details>
<summary>Results (primary 1.0%, secondary 4.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0% | [0.8%, 1.3%] | 2     |
| Regressions ❌ <br /> (secondary)  | 4.1% | [4.1%, 4.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.0% | [0.8%, 1.3%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9200cbc71230a86b8a70481b625d93a4fa5ef523&end=fe43131683267082fc46be5796ce84351e9fb497&stat=cycles:u)
<details>
<summary>Results (secondary 1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9% | [1.9%, 1.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9200cbc71230a86b8a70481b625d93a4fa5ef523&end=fe43131683267082fc46be5796ce84351e9fb497&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.818s -> 781.131s (0.04%)
**Artifact size**: 335.29 MiB -> 335.23 MiB (-0.02%)
